### PR TITLE
Expose Livy Spark UI url and driver log url in %%info magic.

### PR DESCRIPTION
**Pull request by tc0312**
For better debuggability.

### CODE REVIEW

File: remotespark/kernels/kernelmagics.py [hunk 1]
```diff
@@ -147,6 +147,16 @@ def info(self, line, cell="", local_ns=None):
 
         self.ipython_display.writeln("Session configs:\n\t{}\n".format(conf.get_session_properties(self.language)))
 
+        if self.session_started:
+            app_id = self.spark_controller.get_app_id()
+            driver_log_url = self.spark_controller.get_driver_log_url()
+            spark_ui_url = self.spark_controller.get_spark_ui_url()
+        else:
+            app_id = driver_log_url = spark_ui_url = "Session hasn't started."
+        self.ipython_display.writeln("Driver log:\n\t{}\n".format(driver_log_url))
```
Comment (msftristew) [inline remotespark/kernels/kernelmagics.py]: The `\n` at the end isn't needed because this is a call to `writeln` right?

File: remotespark/kernels/kernelmagics.py [hunk 2]
```diff
@@ -147,6 +147,16 @@ def info(self, line, cell="", local_ns=None):
...
+        self.ipython_display.writeln("Spark UI:\n\t{}\n".format(spark_ui_url))
+        self.ipython_display.writeln("YARN Application ID:\n\t{}\n".format(app_id))
```
Comment (msftristew) [inline remotespark/kernels/kernelmagics.py]: I'm unsure about the output here.  The user wants the spark UI link and yarn application id for all sessions, not just the default.

That output should become a part of `LivySession` and included in the `__str__` method for it.

File: remotespark/kernels/kernelmagics.py (hunk 1 above)
Comment (msftristew) [inline remotespark/kernels/kernelmagics.py]: Driver logs shouldn't be gotten here, they should be printed as part of the `%%logs` command, or part of another command (`%%sparklogs`? `%%driverlogs`?)

**Comment by msftristew** [comment]
Spoke with Alex offline, still including URL's in `%%info`, add a second `%%sparkui` magic for opening up a window to the Spark UI.

File: remotespark/livyclientlib/livysession.py [hunk 3]
```diff
@@ -106,11 +107,38 @@ def create_sql_context(self):
         else:
             raise FailedToCreateSqlContextException("Failed to create the SqlContext.\nError: '{}'".format(out))
 
+    def get_app_id(self):
+        if self._app_id is None:
+            session = self._http_client.get_session(self.id)
+            self._app_id = session['appId']
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: Can this field be absent?

Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: Depends on Livy version obviously.  Not a problem for HDInsight but a problem if a person installs this repo and happens to be running against an old Livy version.  Don't know what the implications of this are, we haven't really discussed our approach to Livy versioning.

File: remotespark/livyclientlib/livysession.py [hunk 4]
```diff
@@ -106,11 +107,38 @@ def create_sql_context(self):
...
+    def get_app_info_member(self, member_name, default_value):
+        app_info = self.get_app_info()
+
+        if member_name in app_info and app_info[member_name] is not None:
+            return app_info[member_name]
+        else:
+            return default_value
+
+    def get_driver_log_url(self):
+        return self.get_app_info_member("driverLogUrl", "No driver log URL is available yet.")
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: The message saying that the url is not available yet is misleading for versions of Livy that do not expose the URL at all...

File: remotespark/livyclientlib/livysession.py [hunk 5]
```diff
@@ -106,11 +107,38 @@ def create_sql_context(self):
...
+
     def get_logs(self):
         log_array = self._http_client.get_all_session_logs(self.id)['log']
         self._logs = "\n".join(log_array)
         return self._logs
 
+    def get_spark_ui_url(self):
+        return self.get_app_info_member("sparkUiUrl", "No Spark UI URL is available yet.")
```
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: Ditto

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (aggFTW) [inline remotespark/livyclientlib/livysession.py]: There's no Livy versioning at this moment... What we should do is gracefully handle absences at the very least.

Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: I mean, we do have Livy versioning (a set of Livy versions we support and don't support), it's just not specified.

I would be comfortable with either 1) trying to do this in a backwards-compatible fashion by checking whether the record includes a session ID and 2) pinning the lowest supported version of Livy for publicpreview0.5 to whatever the current version is

File: remotespark/livyclientlib/livysession.py (hunk 4 above)
Comment (tc0312) [inline remotespark/livyclientlib/livysession.py]: Changed it to show "None" like other unavailable fields.

File: remotespark/livyclientlib/livysession.py (hunk 3 above)
Comment (tc0312) [inline remotespark/livyclientlib/livysession.py]: I fixed this and added unit test to make sure things are working if appId/appInfo are missing.

File: remotespark/livyclientlib/livysession.py (hunk 5 above)
Comment (tc0312) [inline remotespark/livyclientlib/livysession.py]: Ditto

**Comment by tc0312** [comment]
Prototyped %%sparkui. Due to its asynchronous nature, browsers built-in popup blocker blocks notebook from opening up Spark UI. Reverting to the old behavior and showing it in %%info.

For long term, we want to make it as a menu item instead of a magic command.

File: remotespark/magics/sparkmagicsbase.py [hunk 6]
```diff
@@ -47,7 +47,7 @@ def _sqlquery(cell, samplemethod, maxrows, samplefraction):
 
     @staticmethod
     def print_endpoint_info(info_sessions):
-        sessions_info = ["        {}".format(i) for i in info_sessions]
+        sessions_info = ["        {}\n".format(i) for i in info_sessions]
```
Comment (aggFTW) [inline remotespark/magics/sparkmagicsbase.py]: Is the new line really necessary?

File: tests/test_command.py [hunk 7]
```diff
@@ -2,10 +2,20 @@
...
 
+def _create_session(kind=SESSION_KIND_SPARK, session_id=-1,
+                    sql_created=False, http_client=None, spark_events=None):
+    if http_client is None:
+        http_client = MagicMock()
+    if spark_events is None:
+        spark_events = MagicMock()
+    ipython_display = MagicMock()
+    session = LivySession(http_client, {"kind": kind}, ipython_display, session_id, sql_created, spark_events)
+    return session
```
Comment (aggFTW) [inline tests/test_command.py]: Why add it here instead of leaving the old method?

File: tests/test_livysession.py [hunk 8]
```diff
@@ -8,7 +8,6 @@
 import remotespark.utils.configuration as conf
 import remotespark.utils.constants as constants
 
-
```
Comment (aggFTW) [inline tests/test_livysession.py]: Please add all removed new lines back so that the file is PEP compliant.

File: tests/test_livysession.py [hunk 9]
```diff
@@ -39,11 +36,18 @@ class TestLivySession(object):
     log_json = json.loads('{"id":6,"from":0,"total":212,"log":["hi","hi"]}')
 
     def __init__(self):
+        self.http_client = None
```
Comment (aggFTW) [inline tests/test_livysession.py]: Why make both http_client and spark_events class instances? I really dislike coupling unit tests like this unless there's an extremely good reason.

File: tests/test_livysession.py [hunk 10]
```diff
@@ -64,15 +68,15 @@ def _next_session_response_post(self, *args):
...
-            http_client = MagicMock()
-        if spark_events is None:
-            spark_events = MagicMock()
+    def _create_session(self, kind=constants.SESSION_KIND_SPARK, session_id=-1, sql_created=False):
         ipython_display = MagicMock()
-        session = LivySession(http_client, {"kind": kind}, ipython_display, session_id, sql_created, spark_events)
+        session = LivySession(self.http_client, {"kind": kind}, ipython_display, session_id, sql_created, self.spark_events)
+        return session
+
+    def _create_session_with_fixed_get_response(self, get_session_json):
```
Comment (aggFTW) [inline tests/test_livysession.py]: All of these helper methods should be `staticmethod`s to ensure needless coupling.

**Comment by aggFTW** [comment]
Could you post a screenshot of formatting for info when notebook has no session vs when it does so that we have an idea/can comment?

File: remotespark/magics/sparkmagicsbase.py (hunk 6 above)
Comment (tc0312) [inline remotespark/magics/sparkmagicsbase.py]: I splitted Spark UI url and driver log url into its own line. Adding a new line between each session make separation between each session clearer.

File: tests/test_livysession.py (hunk 9 above)
Comment (tc0312) [inline tests/test_livysession.py]: http_client and spark_events used by almost all test cases. Moving them to test fixture makes sense.
They are recreated before each test case is executed. There's no coupling between test cases.

File: remotespark/magics/sparkmagicsbase.py (hunk 6 above)
Comment (aggFTW) [inline remotespark/magics/sparkmagicsbase.py]: It seems we need to reformat to an HTML table or similar, as this has gotten messy already? What do you think?

File: tests/test_livysession.py (hunk 10 above)
Comment (tc0312) [inline tests/test_livysession.py]: They rely on test fixture. The coupling is not needless.

File: tests/test_livysession.py (hunk 9 above)
Comment (aggFTW) [inline tests/test_livysession.py]: How would you parallelize these tests with test fixtures like this?

File: remotespark/magics/sparkmagicsbase.py (hunk 6 above)
Comment (tc0312) [inline remotespark/magics/sparkmagicsbase.py]: I agree but I want to get this out first and do it in another commit. Will create an GH issue for it.

Comment (aggFTW) [inline remotespark/magics/sparkmagicsbase.py]: Sounds good!

File: tests/test_livysession.py (hunk 9 above)
Comment (tc0312) [inline tests/test_livysession.py]: By creating test fixture before executing each test case.
Setting _multiprocess_can_split_ = True in the test class will do the trick.

File: tests/test_command.py (hunk 7 above)
Comment (tc0312) [inline tests/test_command.py]: I admit that pasting the old method here isn't the cleanest way. But before my change, it called a member method in another test class, which already looks odd. Ideally we should extract the common method into a base test suite/a test helper class.

**Comment by tc0312** [comment]
image

File: tests/test_livysession.py (hunk 9 above)
Comment (aggFTW) [inline tests/test_livysession.py]: Got it. LGTM.

File: tests/test_command.py (hunk 7 above)
Comment (aggFTW) [inline tests/test_command.py]: True. Do you want to do that?

**Comment by aggFTW** [comment]
We talked f2f.

Remaining items seem to be:
1. Encoding password to be base64 in config file.
2. Making info output an html table. You might want to use: https://ipython.org/ipython-doc/3/config/integrating.html

**Comment by tc0312** [comment]
Will file another PR very soon for 1.
Filed an issue to track [2](https://github.com/jupyter-incubator/sparkmagic/issues/230)

**Comment by msftristew** [comment]
Please resolve conflicts

File: remotespark/livyclientlib/livysession.py [hunk 11]
```diff
@@ -61,7 +62,8 @@ def __init__(self, http_client, properties, ipython_display,
         self.created_sql_context = sql_created
 
     def __str__(self):
-        return "Session id: {}\tKind: {}\tState: {}".format(self.id, self.kind, self.status)
+        return "Session id: {}\tYARN id: {}\tKind: {}\tState: {}\n\tSpark UI: {}\n\tDriver Log: {}"\
```
Comment (msftristew) [inline remotespark/livyclientlib/livysession.py]: This is fine for now but let's close #230 soon.  This is obviously getting unwieldy.

**Comment by msftristew** [comment]
LGTM

**Comment by tc0312** [comment]
Resolved conflicts.

**Comment by msftristew** [comment]
Feel free to merge if it looks good to you, Alejandro
